design/authentication-flow-usability.md

```diff
@@ -0,0 +1,322 @@
+remove just the auth session corresponding to current tab_id and remove root auth session just if there are no more auth sessions (browser tabs).
+This may fix lots of the `You are already logged-in` cases as auth session will be available in the root auth session.
+However it will have impact on memory and at the same time, there will be still cases when the root
+authentication session is expired and hence redirect to the application will be needed.
+
+
+#### Minimal approach - challenges
+
+1. From the user point of view, when the username/password screen is shown in tab2 in step 5 from [above](#the-you-are-already-logged-in-issue),
+the user will be just automatically redirected as authenticated to the application after confirm username/password. Few browser redirects is "implementation detail", which most of the users won't notice.
+However there could be some usability confusions. For example:
+
+    * User will be authenticated even if he provides invalid username/password or even he clicks "Submit" without providing any username/password. As username/password verification didn't really happen in tab2, even if it looks to user as it happened. IMO this is not major issue and same behaviour can be seen by Twitter for example.
+    
+    * After confirm username/password screen, user won't see TOTP (or any other 2 factor authenticators) even if he potentially has 2 factor enabled. He will be directly redirected to the application. I am not sure about usability of this and if it's confusing to users or if it's ok.
+
+2. Another potential challenge is handling of the error on the adapter side.
+
+    * In OAuth2/OIDC, there doesn't seem to be any nice error code for this purpose. See https://openid.net/specs/openid-connect-core-1_0.html#AuthError and https://tools.ietf.org/html/rfc6749#section-4.1.2. It seems best error code might be "server_error" and we may add the custom error_description, so that our adapters can understand it and handle it.
+
+3. In our adapters, we may need to doublecheck if behaviour is correct and user will be redirected back to KC as described in
+the step 3 [above](#minimal-approach). For servlet application, I suppose it won't be an issue as redirect_uri is usually "secured" URI and hence automatically intercepted.
+I think it may be similar for other server adapters (node.js, gatekeeper etc) but will be good to doublecheck.
+For the keycloak.js, I think "login_required" will work fine and automatically redirect to KC, but for the applications with "check_sso" or without any "onLoad" value used,
+(For example _js-console_ application from the deprecated keycloak-examples), the adapter won't automatically redirect to Keycloak. So some changes in the adapter
+may be needed to achieve this.
+
+4. For SAML, will be good to doublecheck the behaviour. I suppose that SAML adapters are server-side and hence is not an
```

Comment:
No issue to redirect back for SAML secured area - the request would be resent unless anonymous access is also allowed into the secured area.

Reply:
So these apps, which allow anonymous access, is exactly my fear. If I understand correctly, after fill username/password in the browser tab2, user might be redirected back to the application, but in the anonymous mode. This can be quite confusing IMO as mentioned in the comment above. If we add some SAML specific errors, adapter can react on them and do the SSO dance instead of showing the anonymous page?

Reply:
I'd rather add such a functionality only if really demanded by the community so that we keep the SAML adapter without Keycloak extensions and stick to the standard codes as much as possible. AFAIK, there is no standard SAML code that we could use.

Reply:
Ok, I am not so concerned about SAML TBH as I don't know how many people use the anonymous mode etc. For OIDC, I would add the error/error_description with some custom values, so that our adapters can react on it instead of showing anonymous page. I think that at least keycloak.js will need some support for this to be added. Simlarly I think that java servlet adapters don't need any change as redirect_uri is always "secured" URL and adapter always redirects to KC when redirect_uri is displayed. Not sure about gatekeeper and node.js